### ISSUE TITLE
JsonSerializerContext not working with minimalApi

### ISSUE THREAD

**Issue by dany28** [dotnet/aspnetcore#52088]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

After configure Web API service to use generated version of `JsonSerializerContext` for serializing JSON, still using default serialization

### Expected Behavior

Object returned by endpoint should be serialized using configured class derived from `JsonSerializerContext` with his configuration. 

### Steps To Reproduce

```C#
using System.Text.Json;
using System.Text.Json.Serialization;

var builder = WebApplication.CreateSlimBuilder(args);

builder.Services.ConfigureHttpJsonOptions(options =>
    options.SerializerOptions.TypeInfoResolverChain.Insert(0, AppJsonSerializerContext.Default));

var app = builder.Build();

app.MapGet("/json", () => new Person("Alex"));
app.MapGet("/string", () => JsonSerializer.Serialize(new Person("Alex"),AppJsonSerializerContext.Default.Person));

app.Run();

record Person(string FirstName)
{
}

[JsonSerializable(typeof(Person))]
[JsonSourceGenerationOptions(PropertyNamingPolicy = JsonKnownNamingPolicy.KebabCaseUpper)]
internal partial class AppJsonSerializerContext : JsonSerializerContext
{

}
```

endpoint `/string` (manually using  AppJsonSerializerContext) returns:
```JSON
{"FIRST-NAME":"Alex"}
```

but endpoint `/json` returns default serialized JSON, without `AppJsonSerializerContext`:
```JSON
{
  "firstName": "Alex"
}
```

### .NET Version

8.0.100

### Anything else?

Project configuration .csproj
```XML
<Project Sdk="Microsoft.NET.Sdk.Web">
    <PropertyGroup>
        <TargetFramework>net8.0</TargetFramework>
        <Nullable>enable</Nullable>
        <ImplicitUsings>enable</ImplicitUsings>
        <InvariantGlobalization>true</InvariantGlobalization>
    </PropertyGroup>
</Project>
```

I've tried debugging generated code, endpoint `/string` goes into generated implementation, but `/json` not

**Comment by ChaosEngine** [comment]
I am seeing the same when NOT providing any `PropertyNamingPolicy` attribute and default json generation (in MVC and minimal API) is misbehaving giving different results. Only proper when explicitly providing `JsonSerializer.Serialize(new MyType() MyTypeContext.Default.MyType)`

Same repro as above except commente out
`//[JsonSourceGenerationOptions(PropertyNamingPolicy = JsonKnownNamingPolicy.KebabCaseUpper)]`

```
using System.Text.Json;
using System.Text.Json.Serialization;

var builder = WebApplication.CreateSlimBuilder(args);

builder.Services.ConfigureHttpJsonOptions(options =>
	options.SerializerOptions.TypeInfoResolverChain.Insert(0, AppJsonSerializerContext.Default));

var app = builder.Build();

app.MapGet("/json", () => new Person("Alex"));
app.MapGet("/string", () => JsonSerializer.Serialize(new Person("Alex"), AppJsonSerializerContext.Default.Person));

app.Run();

record Person(string FirstName)
{
}

[JsonSerializable(typeof(Person))]
//[JsonSourceGenerationOptions(PropertyNamingPolicy = JsonKnownNamingPolicy.KebabCaseUpper)]
internal partial class AppJsonSerializerContext : JsonSerializerContext
{

}
```

expecteed output would be  `{"FirstName":"Alex"}`
while it is `{"firstName":"Alex"}`
(camelCase instead PascalCase

**Comment by dany28** [comment]
Yes, setting PropertyNamingPolicy is not necessary to how JsonSerializerContext is NOT working for minimalApi and MVC (different code for setting)

**Comment by Kumima** [comment]
Same here, when using parameter-binding if you provide the JsonSerializerContext, it will also cause binding fail, which means it does not even use the `JsonSourceGenerationOptions`.
This is really serious, why is there no official feedback?

**Comment by eiriktsarpalis** [comment]
The `JsonSourceGenerationOptions` attribute maps configuration to the `AppJsonSerializerContext .Default` property added by the source generator. The configuration code in the repro is only copying the `TypeInfoResolver` component of the serialization configuration,  in order to also get the desired behavior you would need to write something like
```c#
builder.Services.ConfigureHttpJsonOptions(options
{
    options.SerializerOptions.TypeInfoResolverChain.Insert(0, AppJsonSerializerContext.Default);
    options.SerializerOptions.PropertyNamingPolicy = AppJsonSerializerContext.Default.Options.PropertyNamingPolicy;
});
```
I discussed this with @captainsafia and it seems to me that an accelerator method that copies _all_ configuration from a generated `JsonSerializerContext` to a `JsonOptions` might be warranted (although, N.B. this could create conflicts in applications where more than one generated `JsonSerializerContext` is being consumed).

**Comment by Kumima** [comment]
@eiriktsarpalis So if we have different `PropertyNamingPolicy` for different `JsonSerializerContext`, is there a way to apply multiple Naming Policies automatically based on Type. It will be more convenient, since we've defined the `Policy` through the context.

**Comment by eiriktsarpalis** [comment]
`PropertyNamingPolicy` is a global setting. If you're looking to apply naming policy that is scoped to a particular type or property you should consider using the `JsonPropertyName` attribute instead.

**Comment by Kumima** [comment]
@eiriktsarpalis
```csharp
JsonSerializer.Serialize(new Person("Alex"), AppJsonSerializerContext.Default.Person)
```
Code like above will automaticly resolve the `JsonSourceGenerationOptions` defined through context.
Can we have a way to resolve the `JsonSerializerOptions` automatically for type? It's just weird that we must configure a global `Options` although we've defined that through context.
`JsonSourceGenerationOptions` becomes useless for this scenorio. Some time we don't want a global, and `JsonPropertyNameAttribute` is more verbose than `JsonSourceGenerationOptions`.

**Comment by eiriktsarpalis** [comment]
> Can we have a way to resolve the JsonSerializerOptions automatically for type? It's just weird that we must configure a global Options although we've defined that through context.

But wouldn't you still need to pass a particular context to aspnetcore somehow? These are not tied to the type definition and are not necessarily unique for each type. There's been proposals for type-directed serialization contracts (see https://github.com/dotnet/runtime/issues/90787) but this is more of a long-term vision.

**Comment by Kumima** [comment]
The purpose to provide `JsonSourceGenerationOptions` to context is to set options for group-related objects. They can be grouped by library, functionality, or other purpose. Groups may differ, and global is not appropriate. Adjustment for individual type works, but group configuration brings convenient.
I'm not totally sure why we may separate the `JsonSerializerContext`, isn't it that we need to group related objects?
Configuration like `[JsonSourceGenerationOptions(UseStringEnumConverter = true)]` configures the enum policy for group.

So, I don't think it's totally related to type-directed serialization. And passing a particular context to aspnetcore is like we pass the group configuration, neither global nor individual type.

Anyway, thanks for you explanation, I can understand the work around for now. I just think if `JsonSourceGenerationOptions` works as I expect, it will be more convenient.

**Comment by Kumima** [comment]
@eiriktsarpalis I just remember my main purpose for this issue. The overload method like `httpClient.PostAsJsonAsync("uri", requestObject, AppJsonSerializerContext.Default.RequestObject)` will apply the `JsonSourceGenerationOptions` while asp.net core does not. This brings a huge inconsistency.

**Comment by Kumima** [comment]
https://github.com/dotnet/runtime/issues/107535
This one provides a explanation for this.
